### DISCUSSION TITLE
Repackaging of batik causes huge lag on first use

### DISCUSSION THREAD

**Discussion by Kellindil** [eclipse-orbit/orbit-simrel#31] [General]
Hello ed,

With https://github.com/eclipse-orbit/orbit-simrel/commit/6e25ee9192206677f9d96ff5ff4d62e11d2ea6d8 you've added the "Eclipse-BuddyPolicy:    dependent" header to batik.

Using that version of batik, I'm observing a very long freeze of eclipse on the first use of anything that tries to use SVG files.
When using a sirius (Capella) based product, the first creation of a SVG node in a diagram takes up to 30 seconds because batik now tries to look inside (seemingly) every single jar of the platform to find its services, due to the buddy policy.

Could we have more information on why you added that header? Was there a bug with batik without it?

**Comment by merks** [comment]
These bundles have service registrations that rely on class loader visibility:
```
org.apache.batik.codec_1.17.0.v20231009-1000.jar
        0  2023-08-21 11:37   META-INF/services/
     1503  2023-08-21 11:37   META-INF/services/org.apache.batik.ext.awt.image.spi.ImageWriter
     1290  2023-08-21 11:37   META-INF/services/org.apache.batik.ext.awt.image.spi.RegistryEntry
org.apache.xmlgraphics.source_2.9.0.v20230916-1600.jar
org.apache.xmlgraphics_2.9.0.v20230916-1600.jar
        0  2023-08-21 12:21   META-INF/services/
       48  2023-08-14 13:10   META-INF/services/javax.xml.transform.URIResolver
      273  2023-08-14 13:10   META-INF/services/org.apache.xmlgraphics.image.loader.spi.ImageConverter
      407  2023-08-14 13:10   META-INF/services/org.apache.xmlgraphics.image.loader.spi.ImageLoaderFactory
      448  2023-08-14 13:10   META-INF/services/org.apache.xmlgraphics.image.loader.spi.ImagePreloader
      321  2023-08-14 13:10   META-INF/services/org.apache.xmlgraphics.image.writer.ImageWriter
```
If I recall correctly, it's specifically ImageWriter registered by org.apache.batik.codec that's not found by org.apache.xmlgraphics.
I'm not sure if there were other Class.forName issues as well; it's been a few months...

In principle it's possible to split up the recipes and use https://wiki.eclipse.org/Context_Class_Loader_Enhancements#Eclipse-RegisterBuddy to be more selective, but there needs to be a faster feedback cycle so that I can test that it does not break existing things (BIRT) while improving performance for your use case.  Of course I only expect it to look for services in jars that are downstream of batik, but given the platform is using it, that can be quite a lot of bundles...  I'm not sure what would be special in Capella.  Just bigger and more things downstream from batik?

**Reply by pcdavid** [reply]
GMF Runtime includes some code to initialize this registry without relying on the services API: https://github.com/eclipse-gmf-runtime/gmf-runtime/blob/master/bundles/org.eclipse.gmf.runtime.draw2d.ui.render.awt/src/org/eclipse/gmf/runtime/draw2d/ui/render/awt/internal/Draw2dRenderPlugin.java#L64

**Reply by merks** [reply]
I seems each client finds different workarounds...

I will experiment with the registered buddy thing and see if that also solves the problem but without searching the universe...

If I wrap new versions and provide them in a nightly build, how quickly can they be tested?

**Reply by pcdavid** [reply]
We're testing locally if consuming a version for the Batik 1.17 JARs without the `Eclipse-BuddyPolicy: dependent` fixes the issue in our context. It's using a p2 repo I built locally by removing [this line](https://github.com/eclipse-orbit/orbit-simrel/blob/6e25ee9192206677f9d96ff5ff4d62e11d2ea6d8/maven-bnd/tp/MavenBND.target#L1169).
However this is a locally built subset of Orbit, so not signed, and we can not distribute this as is in our (Capella-based) product.

If we confirm it fixes the issue (or you have an alternative solution), we would like an officially built (and signed) Orbit with the change, so that I can produce a new GMF Runtime version that consumes that (the product consumes Batik via a GMF Runtime repo).

We release our product middle of next week so if you have an alternate solution (or accept to remove the line if we confirm it fixes the issue) before that, we'll try our best to test and integrate it ASAP. That being said, don't sweat it; if we can't fix the issue for this release it will be documented as a known issue and we'll see for a later maintenance version.

**Reply by merks** [reply]
I will look at this with high priority tomorrow first thing then.

**Reply by merks** [reply]
I've changed it to use registered:

https://github.com/eclipse-orbit/orbit-simrel/blob/dd9934865460e6442810fd881642b34e2d6cc18a/maven-bnd/tp/MavenBND.target#L1187-L1188

I've only registered org.apache.batik.awt.util as a buddy, kind of brute force for all batik bundles to keep the recipe simple.   This limits the scope of searching just to batik bundles and only when searching for extensions to org.apache.batik.awt.util's classes.

In the debugger I can see that the service registrations (that GMF doesn't manually) are still discovered automatically with this approach:

![image](https://github.com/eclipse-orbit/orbit-simrel/assets/208716/76cdba28-525d-4be3-9ee1-7b149f79c676)

I expect this will resolve the performance issues you've encountered because of the very limited scope of searching for a very limited subset of cases, i.e., basically just the image registry stuff.

The new versions are available here:

https://download.eclipse.org/tools/orbit/simrel/maven-bnd/nightly/latest
https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/nightly/latest

All have `v20231215-0830` as the qualifier.

**Reply by pcdavid** [reply]
Thanks! We'll try and test ASAP and report back.

**Reply by merks** [reply]
I'm seeing exceptions in BIRT now.  :-(
```
 Caused by: org.apache.batik.transcoder.TranscoderException: Could not write PNG file because no WriteAdapter is availble
	at org.apache.batik.transcoder.image.PNGTranscoder.writeImage(PNGTranscoder.java:122)
	at org.apache.batik.transcoder.image.ImageTranscoder.transcode(ImageTranscoder.java:130)
	... 41 more
```
I'll need to investigate further.

**Reply by merks** [reply]
Yes, there is this _lovely_ code:
```
        WriteAdapter adapter = getWriteAdapter(
                "org.apache.batik.ext.awt.image.codec.png.PNGTranscoderInternalCodecWriteAdapter");
        if (adapter == null) {
            adapter = getWriteAdapter(
                "org.apache.batik.transcoder.image.PNGTranscoderImageIOWriteAdapter");
        }
        if (adapter == null) {
            throw new TranscoderException(
                    "Could not write PNG file because no WriteAdapter is availble");
        }
```
with this implementation for getting the "adapter":
```
    private WriteAdapter getWriteAdapter(String className) {
        WriteAdapter adapter;
        try {
            Class clazz = Class.forName(className);
            adapter = (WriteAdapter)clazz.getDeclaredConstructor().newInstance();
            return adapter;
        } catch (ClassNotFoundException e) {
            return null;
        } catch (InstantiationException e) {
            return null;
        } catch (IllegalAccessException e) {
            return null;
        } catch (NoSuchMethodException e) {
            return null;
        } catch (InvocationTargetException e) {
            return null;
        }
    }
```

**Reply by pcdavid** [reply]
Just for context/reference, Ed is talking about these in Batik itself:
* https://github.com/apache/xmlgraphics-batik/blob/main/batik-transcoder/src/main/java/org/apache/batik/transcoder/image/PNGTranscoder.java#L115
* https://github.com/apache/xmlgraphics-batik/blob/main/batik-transcoder/src/main/java/org/apache/batik/transcoder/image/PNGTranscoder.java#L63

**Reply by pcdavid** [reply]
How does/did BIRT work with previous versions of Batik (up to 1.16) without the custom Eclipse-BuddyPolicy? Do they have an equivalent to [the explicit initialization code from GMF](https://github.com/eclipse-gmf-runtime/gmf-runtime/blob/master/bundles/org.eclipse.gmf.runtime.draw2d.ui.render.awt/src/org/eclipse/gmf/runtime/draw2d/ui/render/awt/internal/Draw2dRenderPlugin.java#L64)?

**Reply by merks** [reply]
I've tested locally that this fixes this specific problem:

https://github.com/eclipse-orbit/orbit-simrel/blob/bf3be88d22fc6abe018dd25df44fa03d69865a48/maven-bnd/tp/MavenBND.target#L1188

New builds are available with qualifier `v20231215-1030`.  I don't expect this change to have any significant performance impact...

**Reply by merks** [reply]
BIRT did some fragment injection things in the past.

**Reply by merks** [reply]
FYI, a BIRT build is running again.  that takes about an hour to complete, if no tests fail earlier...

**Reply by pcdavid** [reply]
Thanks. I'll rebuild a new GMF using `v20231215-1030` for testing on our side.

**Reply by merks** [reply]
Here are all the potentially problematic Class.forName I find

![image](https://github.com/eclipse-orbit/orbit-simrel/assets/208716/7d286c25-2b0c-41e3-a43a-ccaed5f0f297)

**Reply by merks** [reply]
So my sense is that it would be good to add org.apache.batik.svggen and org.apache.batik.dom to prevent problems here, even if those problems are not currently detected by tests.  That's because `org.apache.batik.svggen.DefaultCachedImageHandler.createGraphics(BufferedImage)` does  `Class.forName("org.apache.batik.ext.awt.image.GraphicsUtil")` and `org.apache.batik.dom.ExtensibleDOMImplementation.createCSSEngine(AbstractStylableDocument, CSSContext)` might create  `org.apache.batik.css.parser.ExtendedParserWrapper`.

**Reply by pcdavid** [reply]
It's at https://download.eclipse.org/modeling/gmp/gmf-runtime/updates/interim/I202312151011/
It's on GMF Runtime 1.15.x for the moment as it's the branch our product uses. If it's confirmed to work I'll release both GMF Runtime 1.15.5 and a 1.16.3 (master) with the changed Batik.

**Reply by pcdavid** [reply]
FWIW, in the context of Sirius and Sirius-based modelers (incl. Capella) which use SVG via Batik, we never identified any issue beside the image format registry initialization.

**Reply by merks** [reply]
The BIRT tests are still running.  I made this change:

https://github.com/eclipse-orbit/orbit-simrel/blob/8c4690bab50969f477d9a9bfe13c34512fba3ad1/maven-bnd/tp/MavenBND.target#L1188

The Orbit builds are running and will produce qualifier `v20231215-1130 shortly.  I'll retest BIRT which ran fine with the previous version.  I don't expect new problems.

**Reply by merks** [reply]
Builds are done.   

 The BIRT build is running again:

https://github.com/eclipse-birt/birt/pull/1522

I expect this is final, unless you tell me otherwise.

**Reply by pcdavid** [reply]
We only tested with `v20231215-1030` for the moment, but it fixes the issue in our context, thanks!
I won't have time today, but on monday I'll rebuild a GMF Runtime using the (hopefully) final `v20231215-1130` to check again.

**Reply by ldelaigue** [reply]
Many warm thanks to you @merks for your commitment in having this resolved.

**Reply by merks** [reply]
Thanks for the confirmation guys!  Now I have peace of mind!!   ☺️

**Reply by pcdavid** [reply]
> on monday I'll rebuild a GMF Runtime using the (hopefully) final v20231215-1130 to check again.

It's at https://download.eclipse.org/modeling/gmp/gmf-runtime/updates/milestones/S202312180813/

**Reply by ldelaigue** [reply]
I'm rebuilding against this gmf-runtime milestone for a last check, will keep you posted :)

**Reply by ldelaigue** [reply]
I confirm that the performance degradation we had noticed is no longer there. Thanks guys!

**Reply by merks** [reply]
That's great!  Yeah!!
